runtime/lua/vim/_comment.lua

```diff
@@ -0,0 +1,252 @@
+      indent_width, indent = indent_width_cur, indent_cur
+    end
+
+    -- Update comment info: lines are commented if every line is commented
+    if is_commented then
+      is_commented = comment_check(l)
+    end
+  end
+
+  -- `indent` can still be `nil` in case all `lines` are empty
+  return indent or '', is_commented
+end
+
+---@param x string
+---@return boolean is_blank
+local function is_blank(x)
+  return string.find(x, '^%s*$') ~= nil
+end
+
+---@param parts vim._comment.Parts
+---@param indent string
+---@return fun(line: string): string
+local function make_comment_function(parts, indent)
+  local prefix, nonindent_start, suffix = indent .. parts.left, string.len(indent) + 1, parts.right
+  local blank_comment = indent .. vim.trim(parts.left) .. vim.trim(parts.right)
+
+  return function(line)
+    if is_blank(line) then
+      return blank_comment
+    end
+    return prefix .. string.sub(line, nonindent_start) .. suffix
+  end
+end
+
+---@param parts vim._comment.Parts
+---@return fun(line: string): string
+local function make_uncomment_function(parts)
+  local l_esc, r_esc = vim.pesc(parts.left), vim.pesc(parts.right)
+  local nonblank_regex = '^(%s*)' .. l_esc .. '(.*)' .. r_esc .. '(%s-)$'
+  local blank_regex = '^(%s*)' .. vim.trim(l_esc) .. '(%s*)' .. vim.trim(r_esc) .. '(%s-)$'
+
+  return function(line)
+    -- Try both non-blank and blank regexes
+    local indent, new_line, trail = string.match(line, nonblank_regex)
+    if new_line == nil then
+      indent, new_line, trail = string.match(line, blank_regex)
+    end
+
+    -- Return original if line is not commented
+    if new_line == nil then
+      return line
+    end
+
+    -- Prevent trailing whitespace
+    if is_blank(new_line) then
+      indent, trail = '', ''
+    end
+
+    return indent .. new_line .. trail
+  end
+end
+
+---@param line_start integer
+---@param line_end integer
+---@param ref_position? integer[]
+local function toggle_lines(line_start, line_end, ref_position)
+  ref_position = ref_position or { line_start, 0 }
+  local parts = make_comment_parts(ref_position)
+  local lines = vim.api.nvim_buf_get_lines(0, line_start - 1, line_end, false)
+  local indent, is_comment = get_lines_info(lines, parts)
+
+  local f = is_comment and make_uncomment_function(parts) or make_comment_function(parts, indent)
+
+  -- Direct `nvim_buf_set_lines()` essentially removes both regular and
+  -- extended marks  (squashes to empty range at either side of the region)
+  -- inside region. Use 'lockmarks' to preserve regular marks.
+  -- Preserving extmarks is not a universally good thing to do:
+  -- - Good for non-highlighting in text area extmarks (like showing signs).
+  -- - Debatable for highlighting in text area (like LSP semantic tokens).
+  --   Mostly because it causes flicker as highlighting is preserved during
+  --   comment toggling.
+  _G._from, _G._to, _G._lines = line_start - 1, line_end, vim.tbl_map(f, lines)
```

Comment:
Can we use an exported module variable instead? e.g. `package.loaded['vim._comment']._from`?

Reply:
Sure. Just a bit more hackery needed. As those are guaranteed to be removed later, I thought there is no harm in doing this.

Reply:
I still would prefer us not littering the global scope, just on principle.

Reply:
I'm a follow-up I'll make an internal binding to lockmarks so we can avoid this entirely.

Reply:
I opted for a mix of precomputing command (for numeric values) and using `package.loaded['vim._comment']._lines` for string array. In theory, even `new_lines` can be used inside `string.format()` with `vim.inspect()`, but this is costly and not really reliable.